registry/remote/repository.go

```diff
@@ -943,65 +947,214 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+			resp.Request.URL,
+			err,
+		)
 	}
 
+	// 2. Validate Size
 	size := resp.ContentLength
 	if size == -1 {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: unknown response Content-Length", resp.Request.Method, resp.Request.URL)
+		return ocispec.Descriptor{}, fmt.Errorf(
+			"%s %q: unknown response Content-Length",
+			resp.Request.Method,
+			resp.Request.URL,
+		)
+	}
+
+	// 3. Validate Client Reference
+	var refDigest digest.Digest
+	refDigest, _ = ref.Digest()
+	/*
+		if refDigest, err = ref.Digest(); err != nil {
+			return ocispec.Descriptor{}, fmt.Errorf(
+				"%s %q: invalid reference digest: `%s`; %w",
+				resp.Request.Method,
+				resp.Request.URL,
+				ref,
+				err,
+			)
+		}
+	*/
+
+	// 4. Validate Server Digest (if present)
+	var hdrDigest digest.Digest
+	if hdrDigestStr := resp.Header.Get(dockerContentDigestHeader); hdrDigestStr != "" {
+		if hdrDigest, err = digest.Parse(hdrDigestStr); err != nil {
+			return ocispec.Descriptor{}, fmt.Errorf(
+				"%s %q: invalid response header value `%s`: `%s`; %w",
+				resp.Request.Method,
+				resp.Request.URL,
+				dockerContentDigestHeader,
+				hdrDigestStr,
+				err,
+			)
+		}
 	}
 
-	// validate digest if ref is a digest
-	if refDigest, err := ref.Digest(); err == nil {
-		if err = verifyContentDigest(resp, refDigest); err != nil {
-			return ocispec.Descriptor{}, err
+	/* 5. Now, look for specific error conditions
+
+	This is the Truth Table for when to expect an error, and when not to. This
+	table is intended to be exhaustive and account for all possibilities.
+
+	Anything here with an exclamation mark indicates that the true response is
+	the opposite of what's expected; for example, `!PASS` means we expect a
+	`PASS`, even though it's actually a `FAIL`; that's a function of being
+	bling, and not performing the expensive validation here for performance
+	reasons, knowing that it will be calculated at a later point in time.
+
+	 ___________________________________________________________________________________
+	| ID | CLIENT        | SERVER         | GET                   | HEAD                |
+	|----+---------------+----------------+-----------------------+---------------------+
+	| 1  | tag           | missing        | CALCULATE,PASS        | FAIL                |
+	| 2  | tag           | presentValid   | TRUST,PASS            | TRUST,PASS          |
+	| 3  | tag           | presentInvalid | TRUST,!PASS           | TRUST,!PASS         |
+	| 4  | validDigest   | missing        | CALCULATE,VERIFY,PASS | FAIL                |
+	| 5  | validDigest   | presentValid   | TRUST,COMPARE,PASS    | TRUST,COMPARE,PASS  |
+	| 6  | validDigest   | presentInvalid | TRUST,COMPARE,FAIL    | TRUST,COMPARE,FAIL  |
+	| 7  | invalidDigest | missing        | CALCULATE,VERIFY,FAIL | FAIL                |
+	| 8  | invalidDigest | presentValid   | TRUST,COMPARE,FAIL    | TRUST,COMPARE,FAIL  |
+	| 9  | invalidDigest | presentInvalid | TRUST,COMPARE,!PASS   | TRUST,COMPARE,!PASS |
```

Comment:
"Invalid client digest" is actually treated as tag. Therefore, we don't have case 7-9, do we?

Reply:
Let me see if I understand this on a second look...  Let's look at the case where the server does not have the special header--given that the `ref` is syntactically correct, but otherwise invalid:

1. an "invalid" (syntactically correct digest, but an invalid one) comes in to `generateManifestDescriptor#L1010`
2. `->` [`oras-go:registry/remote/repository.go#L1010`](https://github.com/nima/oras-go/blob/issues/225/registry/remote/repository.go#L1010)
3. `->` [`oras-go:registry/reference.go#L160`](https://github.com/nima/oras-go/blob/issues/225/registry/reference.go#L160)
4. `->` [`go-digest:digest.go#L81`](https://github.com/opencontainers/go-digest/blob/master/digest.go#L81)
5. `->` [`go-digest:digest.go#L103`](https://github.com/opencontainers/go-digest/blob/master/digest.go#L103)
6. `->` [`go-digest:algorithm.go#L179`](https://github.com/opencontainers/go-digest/blob/master/algorithm.go#L179) (returns `nil`, since it's a "valid" digest)
7. `<-` Item `5` returns `nil` (no error) from `Validate`
8. `<-` Item `4` returns `<digest.Digest>, nil` from `Parse`
9. `<-` Item `3` returns `<digest.Digest>, nil` from `Digest`
10. [`oras-go:registry/remote/repository.go#L1011`](https://github.com/nima/oras-go/blob/issues/225/registry/remote/repository.go#L1011) sets `refDigest`
11. [`oras-go:registry/remote/repository.go#L1056`](https://github.com/nima/oras-go/blob/issues/225/registry/remote/repository.go#L1056) returns an error, due to the mismatch, for both `HEAD`, and `GET`

Now, let's look at a tag:

1. *same as above*
2. *same as above*
3. *same as above*
4. *same as above*
5. *same as above*
6. `<-` [`go-digest:digest.go#L107`](https://github.com/opencontainers/go-digest/blob/master/digest.go#L107) (`5`) returns `ErrDigestInvalidFormat`
7. `<-` `4` returns `<digest.Digest>, ErrDigestInvalidFormat` from `Parse`
8. `<-` Item `3` returns `<digest.DigestErrDigestInvalidFormat nil` from `Digest`
9. [`oras-go:registry/remote/repository.go#L1011`](https://github.com/nima/oras-go/blob/issues/225/registry/remote/repository.go#L1011) does not set `refDigest`
10. `HEAD` calls return an error @ [`oras-go:remote/repository.go#L1037`](https://github.com/nima/oras-go/blob/issues/225/registry/remote/repository.go#L1037-L1040)
11. `GET` callers get through unscathed @ [`oras-go:remote/repository.go#L1060`](https://github.com/nima/oras-go/blob/issues/225/registry/remote/repository.go#L1060-L1064)

Those seem different, so I'm trying to understand why you're suggesting that a `tag reference` is equivalent to a "invalid" `digest reference`.

Reply:
Discussion in Slack has resolved this conversation.